core/deployment/src/main/java/io/quarkus/deployment/dev/filesystem/ReloadableFileManager.java

```diff
@@ -0,0 +1,379 @@
+package io.quarkus.deployment.dev.filesystem;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.Vector;
+import java.util.function.Supplier;
+
+import javax.tools.FileObject;
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.StandardLocation;
+
+import io.quarkus.runtime.util.ClassPathUtils;
+
+/**
+ * This reloadable file manager handle the class-paths and file locations of separated file manager instances. <br>
+ * It is required for applications that depends on hot-deployed classes of external sources.
+ *
+ * @see io.quarkus.deployment.configuration.ClassLoadingConfig#reloadableArtifacts
+ *      Quarkus Class Loading Reference
+ */
+public class ReloadableFileManager extends QuarkusFileManager {
+
+    private final StandardJavaFileManager reloadableFileManager;
+
+    public ReloadableFileManager(Supplier<StandardJavaFileManager> supplier, Context context) {
+        this(supplier.get(), supplier.get(), context);
+    }
+
+    protected ReloadableFileManager(StandardJavaFileManager fileManager,
+            StandardJavaFileManager reloadableFileManager, Context context) {
+        super(fileManager, context);
+        this.reloadableFileManager = reloadableFileManager;
+        try {
+            this.reloadableFileManager.setLocation(StandardLocation.CLASS_PATH, context.getReloadableClassPath());
+        } catch (IOException e) {
+            throw new RuntimeException("Cannot initialize reloadable file manager", e);
+        }
+    }
```

Comment:
I found it more efficient to extend from ForwardingJavaFileManager<StandardJavaFileManager> than to implement the StandardJavaFileManager interface manually, the [previous code](https://github.com/quarkusio/quarkus/commit/fcb91468514118cdea9f22d0fb5dcfb6ecbd8498#diff-8a73eeb8360043fb8f5c1b7243bf488395e8e2203c757608fff3eea01d6524b5) of this class is on @aloubyansky branch